### PR TITLE
added castle along with tests

### LINKED ISSUE #3140: [ENH] CASTLE Algorithm
**Is your feature request related to a problem? Please describe.**
CASTLE is a regularization method that improves supervised learning by jointly learning a DAG and a predictive model over the data

**Describe the solution you'd like A clear and concise description of what you want to happen, ideally taking into consideration the existing toolbox design, classes and methods.**
### `_CASTLEModel(nn.Module)` (Internal)
PyTorch masked autoencoder for feature reconstruction, target prediction, and training.

- **`__init__(num_inputs, model_cfg: ModelConfig, reg_cfg: RegularizationConfig)`**: Initializes masked input layers, scalar output layers, grouped hyperparameters, and optimizer.
- **`forward(X)`**: Returns full reconstruction (`Out`) and target prediction (`out_0`).
- **`train(X_tensor)`**: Runs epochs with early stopping (batch updates, Lagrangian multiplier adjustments), logs to TensorBoard if enabled, and returns the thresholded adjacency matrix (`W_final`).
- **`get_W()`**: Computes adjacency matrix from current weights (L2 norm of column $j$ in sub-network $i$).

### Internal dataclasses
Grouped configuration objects used by `_CASTLEModel`.

- **`ModelConfig`**: `hidden_dim`, `batch_size`, `max_epochs`, `optimizer`, `seed`, `min_loss_improvement`, `early_stop_patience`, `tensorboard_log_dir`, `scaler`, `target_col`
- **`RegularizationConfig`**: `dag_weight`, `sparsity_weight`, `dag_penalty`, `edge_threshold`

### `CASTLE(_BaseCausalDiscovery)` (Public API)
Validates data, orchestrates training, and builds the `pgmpy.DAG`.

```python
CASTLE( 
    dag_weight=1.0,
    sparsity_weight=5.0,
    dag_penalty=1.0,
    optimizer=None,
    batch_size=32,
    hidden_dim=32,
    edge_threshold=0.3,
    target_col=None,
    max_epochs=200, 
    min_loss_improvement=1e-4,
    early_stop_patience=10,
    scaler=None,
    tensorboard_log_dir=None,
    seed=None,
)
```

Joint training objective being minimized:

$$\min_\Theta \frac{1}{N}\|Y - [f_\Theta(\tilde{X})]_{:,1}\|^2 + \lambda \underbrace{\left(L_N(f_\Theta) + (\text{tr}(e^{M \odot M}) - d - 1)^2 + \beta V_{\Theta_1}\right)}_{R_{\text{DAG}}}$$

- `dag_weight` (λ): Weight on the entire DAG regularization loss $R_{\text{DAG}}$.
- `sparsity_weight` (β): Group-lasso penalty on input-layer weights $V_{\Theta_1}$, promoting edge sparsity.
- `dag_penalty` (ρ): Initial penalty coefficient in the Augmented Lagrangian for the acyclicity constraint.
- `optimizer`: Any `torch.optim.Optimizer` instance. Defaults to `Adam(lr=1e-3)` if `None`. Note: `weight_decay` is not recommended when passing a custom optimizer as CASTLE has its own sparsity mechanism via `dag_weight` and `sparsity_weight`.
- `batch_size`: Number of samples per mini-batch.
- `hidden_dim` (h): Hidden layer width in each sub-network $f_k$.
- `edge_threshold`: Edges with weight below this value are zeroed in the final DAG.
- `target_col`: Target column name or index. Defaults to the first column when `None`.
- `max_epochs`: Maximum number of training epochs.
- `min_loss_improvement`: Minimum decrease in total loss per epoch to count as an improvement for early stopping.
- `early_stop_patience`: Number of consecutive epochs without improvement before stopping early.
- `scaler`: Optional scaler for input normalization (for example, `sklearn.preprocessing.StandardScaler`). It must implement `fit(X)` and `transform(X)`; `inverse_transform(X)` is optional for user-side de-scaling. If provided, it is fit on training data and stored as `scaler_`.
- `tensorboard_log_dir`: If set, logs training metrics via `SummaryWriter(log_dir=...)`. If `None`, TensorBoard logging is disabled.
- `seed`: Seed for reproducibility.

**Methods:**
- **`fit(X)`**: Trains the model and builds the DAG.

**Attributes set after `fit`:**
- `n_features_in_`: Number of input features seen during fit.
- `feature_names_in_`: Feature names from the input DataFrame.
- `causal_graph_`: Learned causal DAG as a `pgmpy.base.DAG`.
- `adjacency_matrix_`: Weighted adjacency matrix as a Pandas DataFrame.
- `model_`: Internal neural network instance used for training.
- `scaler_`: Fitted scaler used to normalize inputs during training (when provided).

---
## Usage

```python
import torch
from pgmpy.causal_discovery import CASTLE
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler

df = pd.DataFrame(np.random.randn(100, 3), columns=['A', 'B', 'Target'])

# Default: Adam with lr=1e-3
model = CASTLE(max_epochs=200, edge_threshold=0.3, seed=42)

# Custom optimizer — user controls all optimizer params directly
opt = torch.optim.Adam(lr=1e-4, weight_decay=1e-5)
model = CASTLE(max_epochs=200, edge_threshold=0.3, optimizer=opt, seed=42)

# User-provided scaler and early stopping defaults
scaler = StandardScaler()
model = CASTLE(
    max_epochs=200,
    edge_threshold=0.3,
    target_col="Target",
    scaler=scaler,
    min_loss_improvement=1e-4,
    early_stop_patience=10,
    tensorboard_log_dir="runs/castle",
    seed=42,
)
model.fit(df)
dag = model.causal_graph_

# User-side scaling and inverse-scaling for predictions (example)
scaler_fitted = model.scaler_
X_scaled = scaler_fitted.transform(df.values)
y_pred_scaled = some_model(X_scaled)
y_pred = scaler_fitted.inverse_transform(y_pred_scaled)
```

View logs with: `tensorboard --logdir runs/castle`

**Additional context**
paper: https://arxiv.org/pdf/2009.13180
ref codebase: https://github.com/trentkyono/CASTLE

Description: related : #2931, pgmpy/enhancement_proposals#4

Note this issue is part of an ongoing mentorship project

### PR DESCRIPTION
**Pull request by Manas-7854**
minor implementation details:
- currently using Standar Scalar, can allow change that to allowing user to changing the scalar method
- earlier implementation had early stopping but since the  implementation provided in the paper did not have early stopping that was removed, can add that as optional 
- the paper's goal is to learn a DAG via continuous optimisation, not to guarantee one, thus the dag we are returning need not to be perfect, as of now if the obtained dag is not acyclic, there are not warning.(can add a warning for that if required)

**The following checklist is mandatory**

Your PR will be closed if you remove the checklist. Use of LLMs is **strictly forbidden** for any part of this checklist (including for improving language), and will result in a **ban** if we find any use of LLMs.

### Your checklist for this pull request

- [x] Have you followed all the steps from our [Contributing Guide](https://github.com/pgmpy/pgmpy/blob/dev/Contributing.md)?
- [x] Does the PR fully address the linked issue and is within its defined scope? If you are still working on the PR, mark it as draft.
- [x] Are all the GitHub Actions checks passing? If not, mark your PR as draft while you fix it.

If you have used AI/LLMs for any assistance, please answer the following questions. Please refer [#2622](https://github.com/pgmpy/pgmpy/pull/2622) for an example of the level of detail we expect:

- [x] Have you reviewed our [AI usage policy](https://github.com/pgmpy/pgmpy/blob/dev/Contributing.md#ai-usage-policy)?
- [x] Are you able to fully explain your changes? We expect you to fully understand the algorithm and take full responsibility for any changes in this PR.  

- Please describe in **detail** how and what you used AI assistance for? Please outline your whole workflow with the AI tool.

Claude was used to implement parts of PR, I started by reading the paper and understanding the algorithm and coming up with a detailed implementation plan along with design choices. Description of all the implemented functions and classes was given to Claude, based on that an initial draft of code was provided by Claude. After that I did a thorough review of the code, verified if it it follows the design from the paper exactly and had to make a lot of minor changes and bug fixes.
After this I gave a list of test cases to Claude to provide me with the code for them. The test cases were designed by me
1. some inspiration for the test cases was taken from the codebase in the, which has an implementatoin of CASTLE algorithm
2. a lot of the test cases were checks on specific parts of algorithm
3. i came up with some of that test cases to test the final prediction of the model
Given a the code for the test cases a thorough review was done and minor changes were made on both `castle.py` and `test_castle.py`

- What steps have you taken to verify that the changes correctly address the issue? What edge cases have you considered? Other than running tests, what else have you verified?  

the most important verification I did was, going through the paper in depth again and again and making sure everything was implemented based on the paper. Initial implementation had generating a DAG as the primary objective however the paper makes it clear that the supervised prediction is the primary objective and the R_DAG is the regularizer.
Edge cases consider: 
- w_threshold = 0 , all weight kep, dag constraint may not have converged, potential cycles. Handled with a user warning rather that silently modifying the graph
- w-thresholds very high- produces an empty graph, which is a valid dag

- Have you used AI for generating tests? Can you compress them into a smaller number of tests without losing coverage?

The AI was given a detailed list of test cases to implement, which I had come up with. The AI just provided me with the code, which was further verified by me. No the test cannot be compresses into smaller number without losing coverage

### Issue number(s) that this pull request fixes
- Fixes #3140 

### List of changes to the codebase in this pull request
- added `pgmpy/causal_discovery/castle.py`
- added `pgmpy/tests/test_causal_discovery/test_CASTLE.py`
- modified `pgmpy/causal_disccover/_init_.py` by importing Castle to, also added a `_gettatr_` functiont to check for soft dependencies to avoid dependency issues

Implemented the Castle Algorithm: CASTLE is a regularization method that improves supervised learning by jointly learning a DAG and a predictive model over the data

### CODE REVIEW

File: pgmpy/causal_discovery/castle.py [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+        self.adjacency_matrix_ = pd.DataFrame(W_final, columns=cols, index=cols)
+
+        dag = DAG()
+        dag.add_nodes_from(cols)
+        for i in range(num_inputs):
+            for j in range(num_inputs):
+                if W_final[j, i] != 0.0:
+                    dag.add_edge(cols[j], cols[i])
+
+        self.causal_graph_ = dag
```
[the remark this answers is not in the document]
Comment (Manas-7854) [reply pgmpy/causal_discovery/castle.py:367]: The CASTLE paper's goal is to learn a DAG via continuous optimisation, not to guarantee one, thus the dag we are returning need not to be perfect

File: pgmpy/causal_discovery/CASTLE.py [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+        Returns
+        -------
+        self : CASTLE
+        """
+        cols = list(X.columns)
+
+        if target_col is None:
+            target_col = cols[0]
+        elif target_col not in cols:
+            if isinstance(target_col, int) and target_col < len(cols):
```
[the remark this answers is not in the document]
Comment (Manas-7854) [reply pgmpy/causal_discovery/CASTLE.py]: fixed this

File: pgmpy/tests/test_causal_discovery/test_castle.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+
+import numpy as np
+import pandas as pd
+import pytest
+import networkx as nx
+from unittest.mock import patch
+
+from pgmpy.base import DAG
+from pgmpy.causal_discovery import CASTLE
+
```
[the remark this answers is not in the document]
Comment (Manas-7854) [reply pgmpy/tests/test_causal_discovery/test_castle.py:17]: fixed this by adding `check_soft_dependencies` in test file as well

File: pgmpy/causal_discovery/castle.py [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+        self.model_ = model
+        self.cols_ = cols
+        self.feature_names_in_ = cols[1:]
+
```
[the remark this answers is not in the document]
Comment (Manas-7854) [reply pgmpy/causal_discovery/castle.py:371]: fixed this

**Comment by Manas-7854** [comment]
hi @ankurankan , since i am using `torch` inside `castle.py`, i was running a lot of dependency issues and a lot of checks were failing, to fix that i have added a `_gettatr_` inside of `init.py`, as of now the all the checks are passing, however i am not sure this is the best way to fix this error.

Also is there a way i can run specific checks on my system before commiting, the code is passing the pre-commit checks was failing on a few checks after pushing.

PS: apologies for soo many commits

**Comment by ankurankan** [comment]
@Manas-7854 Please add the full PR checklist back.

**Comment by ankurankan** [comment]
@Manas-7854 I would also suggest discussing the design first on the issue before implementing it.

**Comment by Manas-7854** [comment]
Apologies @ankurankan,
- I have added back the PR checklist

**Comment by Manas-7854** [comment]
#### Here are the design choices I have made, let me know if any of these decisions need to be revisited before continuing with the PR review:
- **API- structed fits the existing `_BaseCausalDiscovery` pattern**: `CASTLE` subclasses `_BaseCausalDiscovery`  with a public fit that calls `_check_fit_data` then delegates to `_fit` matching the implementation of current `GES` and `PC`.
- `**targel_col` lives in `fit`, not `__init__**`: the paper treats target column as data-level concern as model hyperpapermeter, thus puting it inside `__init__` would mean re-instantiating the model to change the target, which felt wrong
- **Torch is a soft dependency**: since torch is not in pgmpy's core dependencies, the gaurd follows the existing pgmpy's patter to user `_check_soft _dependency` to handle this dependency, also `_safe_import` is used at module level so that import `pgmpy.causal_discovery` never fails even without torch. `_CASTLEModel` and `_dag_constraint` are wrapped in `if torch is not None` gaurd with light weight placeholder classes int he `else` branch.
- **Augmented Lagrangian schedule for the DAG penalty:** `rho` and `alpha` are updated at the epoch level, not the batch level, following `notears` augmented lagrangian convention
- **DAG constraint uses** : `torch.linalg.matrix_exp`, directly rather than a truncted power series, which is more numerically stable and removes the arbitrary truncation at k=10
- since the primary objective in the paper was to use DAG as a regularizer for supervised prediction the might DAG after the implementation might not be perfect

cc : @ankurankan

**Comment by ankurankan** [comment]
@Manas-7854 I would like to review this live on a call. Can you send me a message on discord to figure out a time?

**Comment by ankurankan** [comment]
@Manas-7854, I have put some more thought into these causal discovery algorithms with a target variable. I had completely missed that we also already had tree search-based algorithms that require a target node. So, I think we should go ahead with adding this algorithm to causal discovery. I do think we should be able to also use these algorithms for feature selection kind of applications.

Let me know when this is ready for another round of review.

**Comment by Manas-7854** [comment]
@ankurankan that sounds good! Do you want to continue the review on a call?

**Comment by ankurankan** [comment]
@Manas-7854 Is this ready for another round of review? Did we discuss some changes on our last call?
